aidial_adapter_vertexai/embedding/multi_modal.py

```diff
@@ -0,0 +1,224 @@
+        )
+
+    data = await download_attachment(file_storage, attachment)
+    return Image(image_bytes=data)
+
+
+async def get_requests(
+    request: EmbeddingsRequest,
+    storage: FileStorage | None,
+) -> AsyncIterator[ModelRequest]:
+    async def on_text(text: str):
+        return ModelRequest(contextual_text=text)
+
+    async def on_attachment(attachment: Attachment):
+        return ModelRequest(image=await download_image(storage, attachment))
+
+    async def on_mixed(inputs: List[str | Attachment]) -> ModelRequest:
+        if len(inputs) == 0:
+            raise EMPTY_INPUT_LIST_ERROR
+        elif len(inputs) == 1:
+            if isinstance(inputs[0], str):
+                return await on_text(inputs[0])
+            else:
+                return await on_attachment(inputs[0])
+        elif len(inputs) == 2:
+            if isinstance(inputs[0], str) and isinstance(inputs[1], Attachment):
+                return ModelRequest(
+                    contextual_text=inputs[0],
+                    image=await download_image(storage, inputs[1]),
+                )
+            elif isinstance(inputs[0], Attachment) and isinstance(
+                inputs[1], str
+            ):
+                return ModelRequest(
+                    contextual_text=inputs[1],
+                    image=await download_image(storage, inputs[0]),
+                )
+            else:
+                raise ValidationError(
+                    "The first element of a custom_input list element must be a string and the second element must be an image attachment or vice versa"
+                )
+        else:
+            raise ValidationError(
+                "No more than two elements are allowed in an element of custom_input list"
+            )
+
+    return collect_embedding_inputs(
+        request,
+        on_text=on_text,
+        on_attachment=on_attachment,
+        on_mixed=on_mixed,
+    )
+
+
+class MultiModalEmbeddingsAdapter(EmbeddingsAdapter):
+    model_id: str
+    model: MultiModalEmbeddingModel
+    api_key: str
+    storage: FileStorage | None
+
+    @classmethod
+    async def create(
+        cls,
+        model_id: str,
+        project_id: str,
+        location: str,
+        api_key: str,
+    ) -> "EmbeddingsAdapter":
+        storage = create_file_storage(api_key)
+        await init_vertex_ai(project_id, location)
+        model = await get_multi_modal_embedding_model(model_id)
+        return cls(
+            model_id=model_id, model=model, api_key=api_key, storage=storage
+        )
+
+    async def embeddings(
+        self, request: EmbeddingsRequest
+    ) -> EmbeddingsResponse:
+
+        validate_request(request)
+
+        vectors: List[List[float] | str] = []
+        token_count = 0
+
+        # NOTE: The model doesn't support batched inputs
+        async for sub_request in await get_requests(request, self.storage):
```

Comment:
Yeah,  model doesn't support batches, but we can still speed this up drastically, if we  use `asyncio.gather` instead of `for loop`

Look at this test script
```python
import asyncio
import time
from concurrent.futures import ThreadPoolExecutor
import io
from PIL import Image as PILImage
import random

from vertexai.vision_models import MultiModalEmbeddingModel, Image

model = MultiModalEmbeddingModel.from_pretrained("multimodalembedding@001")


def add_salt(image: PILImage.Image, salt_factor=0.01):
    img_array = image.copy().convert("RGB")
    width, height = img_array.size
    pixels = img_array.load()

    for _ in range(int(width * height * salt_factor)):
        x = random.randint(0, width - 1)
        y = random.randint(0, height - 1)
        pixels[x, y] = (
            random.randint(0, 255),
            random.randint(0, 255),
            random.randint(0, 255),
        )

    img_byte_arr = io.BytesIO()
    img_array.save(img_byte_arr, format="PNG")
    return Image(image_bytes=img_byte_arr.getvalue())


def get_embedding(image, text):
    return model.get_embeddings(image=image, contextual_text=text)


async def async_get_embedding(loop, executor, ri):
    image, text = ri
    return await loop.run_in_executor(executor, get_embedding, image, text)


def sequential_calls(randomized_inputs):
    start_time = time.time()
    for ri in randomized_inputs:
        image, text = ri
        get_embedding(image, text)
    return time.time() - start_time


async def async_calls(randomized_inputs):
    start_time = time.time()
    tasks = []
    loop = asyncio.get_event_loop()
    with ThreadPoolExecutor() as executor:
        for ri in randomized_inputs:
            task = async_get_embedding(loop, executor, ri)
            tasks.append(task)
        await asyncio.gather(*tasks)
    return time.time() - start_time


async def main():
    IMAGES_COUNT = 10
    with open("./sample_image.png", "rb") as file:
        original_image = PILImage.open(io.BytesIO(file.read()))

    # ====== SYNC CALLS
    print("Testing sequential calls:")
    randomized_inputs = [
        (add_salt(original_image), f"Some contextual text {index}")
        for index in range(IMAGES_COUNT)
    ]
    print("Randomized images for sync are prepared")
    sequential_time = sequential_calls(randomized_inputs)
    print(f"Sequential time: {sequential_time:.2f} seconds")

    # ====== ASYNC CALLS
    print("=" * 30, "\n\n\nTesting async calls:")
    async_randomized_inputs = [
        (add_salt(original_image), f"Some contextual text {index}")
        for index in range(IMAGES_COUNT)
    ]
    print("Randomized images for async are prepared")
    async_new_executor_time = await async_calls(async_randomized_inputs)
    print(f"Async time: {async_new_executor_time:.2f} seconds")


if __name__ == "__main__":
    asyncio.run(main())
```

Reply:
In this test I got a substantial speed up on 10 images:
```
Testing sequential calls:
Randomized images for sync are prepared
Sequential time: 15.17 seconds

============================== 


Testing async calls:
Randomized images for async are prepared
Async time: 2.43 seconds
```

Reply:
Fixed